text/2023-04-26-color-outlines.md

```diff
@@ -0,0 +1,198 @@
+```
+This is guaranteed to be the first method called on the pen to allow an
+appropriately sized surface to be allocated if necessary. It will pass the
+value of the `ClipBox` if available, otherwise a bounding box computed
+during evaluation of the paint graph.
+
+In accordance with the spec, if the paint graph is determined to be
+unbounded, no methods on the pen will be called and the `color_outline`
+method will return an error.
+
+```rust
+fn push_transform(&mut self, transform: Transform)
+```
+Push a transform onto the state stack.
+
+> `canvas->save();`<br>
+`canvas->concat(TransformToSkMatrix(transform));`
+
+```rust
+fn pop_transform(&mut self)
+```
+Pop a transform from the state stack.
+
+> `canvas->restore();`<br>
+
+```rust
+fn push_clip(&mut self, glyph_id: Glyph_id, path: &Path)
+```
+Push a clip path onto the state stack.
+
+> `SkPath clipPath = PathToSkPath(path);`<br>
+`canvas->save();`<br>
+`canvas->clip(clipPath, true);`
+
+```rust
+fn pop_clip(&mut self)
+```
+Pop a clip path from the state stack.
+
+> `canvas->restore();`<br>
+
+```rust
+fn push_layer(&mut self, mode: Option<CompositeMode>)
+```
+Push a new layer to the state stack, optionally with a composite mode.
+
+If `mode` is not `None`:
+> `SkPaint blendModePaint;`<br>
+`blendModePaint.setBlendMode(CompositeModeToSkBlendMode(mode.unwrap());`<br>
+`canvas->saveLayer(nullptr, &blendModePaint);`
+
+Otherwise:
+> `canvas->saveLayer(nullptr, nullptr);`<br>
+
+```rust
+fn pop_layer(&mut self)
+```
+Pop a layer from the state stack.
+
+> `canvas->restore();`<br>
+
+```rust
+fn fill(&mut self, brush: &Brush);
+```
+Fill the current clip with the given brush.
+
+> `canvas->drawPaint(BrushToSkPaint(brush));`<br>
+
+```rust
+fn fill_path(
+    &mut self,
+    glyph_id: GlyphId,
+    path: &Path,
+    brush: &Brush,
+    brush_transform: Option<Transform>
+)
+```
+Fill the specified path with the given brush. Note that this is an optimization
+for the pattern `Clip Transform* Fill` in the graph. The intermediate
+transforms are captured in the optional `brush_transform` parameter.
+
+Assuming `brush_transform` is `None`:
+> `SkPath fillPath = PathToSkPath(path);`<br>
+`SkPaint fillPaint = BrushToSkPaint(brush);`<br>
+`canvas->drawPath(fillPath, fillPaint);`<br>
+
+Otherwise, use the standard clip, transform, fill code.
+
+## Questions
+
+### Should transforms be absolute?
+
+The transform stack is generally unnecessary after flattening the graph. We can
+remove the `push_transform`/`pop_transform` methods and provide absolute 
+transforms to the methods where they are required.
+
+### How much processing on gradient fills?
+
+Skia does a significant amount of
+[processing](https://skia.googlesource.com/skia/+/265a074c910a0a4ac29176bca0a34615247f10f7/src/ports/SkFontHost_FreeType_common.cpp#597)
+on gradient fills to match the expectations of the relevant shaders. The
+applied transformations appear to produce equivalent brushes with color stops
```

Comment:
As discussed, my suspicion here is that other 2D libs have different expectations, in particular for the sweep type of gradients. And while the output is hopefully really equivalent, it may then introduce two layers of value conversion (first in Skrifa, then on the client side) when using this functionality with other 2D libraries. Perhaps we can at least look at Cairo and Direct2D to get a sense for that.